Add Posix netinet_in.h additions

jukkar opened the pull request:
Most of the tests in `tests/posix/headers` related to networking are already supported so mark them so.
Add implementation to various IPv6 address checking macros we were missing and tests for them.

jukkar committed:
tests: posix: headers: net: sin6_scope_id is supported

The Posix documentation uses sin6_scope_id for scope id and
we support that already.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
tests: posix: headers: net: Add check for in6addr_any variable

The in6addr_any is a global symbol that can be used when any
address (all zero bits) is being referenced.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
tests: posix: headers: net: Add IPv6 join/leave group

The IPV6_JOIN/LEAVE_GROUP corresponds to IPV6_ADD/DROP_MEMBERSHIP
definitions so define the former and add tests for them.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
tests: posix: headers: net: Mark unicast/multicast hops supported

Both the IPV6_UNICAST_HOPS and IPV6_MULTICAST_HOPS are supported
so add tests for them.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
tests: posix: headers: net: Add INADDR_BROADCAST definition

Add INADDR_BROADCAST definition and a test for it.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
tests: posix: headers: net: Add various IPv6 address check macros

Add these Posix specified IPv6 address checking macros

IN6_IS_ADDR_UNSPECIFIED
IN6_IS_ADDR_LOOPBACK
IN6_IS_ADDR_MULTICAST
IN6_IS_ADDR_LINKLOCAL
IN6_IS_ADDR_SITELOCAL
IN6_IS_ADDR_V4MAPPED
IN6_IS_ADDR_MC_NODELOCAL
IN6_IS_ADDR_MC_LINKLOCAL
IN6_IS_ADDR_MC_SITELOCAL
IN6_IS_ADDR_MC_ORGLOCAL
IN6_IS_ADDR_MC_GLOBAL

and tests for them.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
tests: posix: headers: net: Add test for ipv6_mreq

The "struct ipv6_mreq" is supported so enable the tests for it.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

cfriedt reviewed (dismissed):
LGTM

For reference -

https://pubs.opengroup.org/onlinepubs/9799919799/basedefs/netinet_in.h.html

pdgendt commented on tests/posix/headers/src/netinet_in_h.c:
```diff
@@ -58,31 +73,31 @@ ZTEST(posix_headers, test_netinet_in_h)
 
 	zassert_not_equal(-1, INADDR_ANY);
-	/* zassert_not_equal(-1, INADDR_BROADCAST); */ /* not implemented */
+	zassert_equal(0xffffffff, INADDR_BROADCAST);
```
I'd recommend using `zexpect_equal` here (and other tests too). This will show if multiple tests would fail, rather than if only the first one.

jukkar replied:
That is a great point, I will fix the tests. I think this pr will go to 4.1, no point trying to get it to 4.0.

jukkar committed:
tests: posix: headers: net: Convert to use zexpect

Use zexpect instead of zassert so that the failed tests do not
stop the test execution but we run all the tests.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar commented:
* based against latest main
* converted tests to use zexpect instead of zassert
